src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,89 @@
-		*dim_input_space2 = 0;
-		*kernelwidth2=1;
-		*randomcoeff_additive2 = NULL;
-		*randomcoeff_multiplicative2 = NULL;
-		return;
-	}
-
-	*dim_feature_space2 = cur_dim_feature_space;
-	*dim_input_space2 = cur_dim_input_space;
-	*kernelwidth2=cur_kernelwidth;
-
-	*randomcoeff_additive2 = SG_MALLOC(float64_t, cur_dim_feature_space);
-	*randomcoeff_multiplicative2 = SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive, randomcoeff_additive+cur_dim_feature_space,
-			*randomcoeff_additive2);
-	std::copy(randomcoeff_multiplicative, randomcoeff_multiplicative+cur_dim_feature_space*cur_dim_input_space,
-			*randomcoeff_multiplicative2);
-
-
-}
-
-void RandomFourierGaussPreproc::set_randomcoefficients(
-		float64_t *randomcoeff_additive2,
-		float64_t * randomcoeff_multiplicative2,
-		const int32_t dim_feature_space2, const int32_t dim_input_space2, const float64_t kernelwidth2) {
-	dim_feature_space = dim_feature_space2;
-	dim_input_space = dim_input_space2;
-	kernelwidth=kernelwidth2;
-
-	SG_FREE(randomcoeff_multiplicative);
-	randomcoeff_multiplicative=NULL;
-	SG_FREE(randomcoeff_additive);
-	randomcoeff_additive=NULL;
-
-	cur_dim_feature_space=dim_feature_space;
-	cur_dim_input_space = dim_input_space;
-	cur_kernelwidth=kernelwidth;
-
-	if( (dim_feature_space>0) && (dim_input_space>0) )
-	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
-	}
+	UniformRealDistribution<float64_t> uniform(0.0, 2.0 * Math::PI);
 
+	for (auto row : range(m_dim_output))
+		m_offset[row] = uniform(m_prng);
 }
 
 void RandomFourierGaussPreproc::fit(std::shared_ptr<Features> f)
 {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"CRandomFourierGaussPreproc::init (Features *f): dim_feature_space<=0 is not allowed, use void set_dim_feature_space(const int32 dim) before!\n");
-	}
-
-	io::info("calling CRandomFourierGaussPreproc::init(...)");
-	int32_t num_features =
-	    f->as<DenseFeatures<float64_t>>()->get_num_features();
-
-	if (!test_rfinited()) {
-		dim_input_space = num_features;
-		init_randomcoefficients();
-		ASSERT( test_rfinited())
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	auto num_features = f->as<DenseFeatures<float64_t>>()->get_num_features();
+	require(num_features > 0, "Input space dimension must be greater than zero");
+	
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
```

Comment:
since you are refactoring this can you change the signature to `const SGVector<float64_t>&` please?

Reply:
this signature is inherited, would have to change a lot of files for it to work. thoughts?

Reply:
hmm I guess let it be for another PR then :)

Reply:
yes, there will be a follow up refactoring the whole class structure.